### PR TITLE
Move exception handling logic to Route

### LINKED ISSUE #493: Question about Request.Body()
Hi dears developers, how can I retrieve the request.body inside an exception_handler? 

My scenario:

I have an exception_handler hook that receives a request and an exception, I'm also using FastApi which tries to parse the body to a JSON, if it does not work it raises an HTTPException. So I'm trying to build a hook to log invalid JSON requests, this exception is being correctly captured by my exception_handler but when I'm trying to retrieve the body I can't it hangs, it simple doesn't work. It does make sense as after reading for the first time it is being cached inside the Request object but the Request object is being recreated before calling the exception_handler so I'm losing my cache and I'm able to ready from the stream again.

https://github.com/encode/starlette/blob/d23bfd0d8ff68d535d0283aa4099e5055da88bb9/starlette/exceptions.py#L79

Not sure if it is supposed to be like this by design, but I think my scenario is not that fancy, it should be something possible, what do you guys think? Any workaround?

My code:

```@app.exception_handler(HTTPException)  
async def httpexception_logger(request: Request, exc: HTTPException):
    body = await request.body()
    text: str = bytes.decode(body)
    logger.error("Exception at %s, status code: %s, body: %s: exception %s", request.url, exception.status_code, text, exc)
    ... other stuff...
```

### PR DESCRIPTION
**Pull request by adriangb**
Fixes #493, partially resolves https://github.com/encode/starlette/pull/1692#issuecomment-1266401636, resolves https://github.com/encode/starlette/pull/1649#issuecomment-1404473365

By moving handling of exceptions to the `Route` level we can use the same `Request`/`WebSocket` object for the endpoint and exception handlers.

See #2020 for some more background

### CODE REVIEW

**Comment by Kludex** [comment]
I'm positive towards this PR. Is any other member that wants to check it? @encode/maintainers

**Comment by adriangb** [comment]
@tomchristie it would be great to get your eyes on this since no one has really changed this part of the codebase since you originally wrote it

File: starlette/middleware/exceptions/_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from starlette._utils import is_async_callable
+from starlette.concurrency import run_in_threadpool
+from starlette.exceptions import HTTPException
+from starlette.requests import Request
+from starlette.responses import Response
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
+from starlette.websockets import WebSocket
+
+Handler = typing.Callable[..., typing.Any]
+ExcHandlers = typing.Dict[typing.Any, Handler]
```
Comment (alex-oleshkevich) [inline starlette/middleware/exceptions/_wrapper.py]: Let's use a full version here: `ExceptionHandlers`

**Review by Kludex — COMMENTED** [review]
Are you strong about this code structure?

Do you mind if we maintain the `exceptions.py` and create a `_<some_good_name_at_root_level>`? I feel like importing `wrap_app_handling_exceptions` from `middleware/exceptions` is not very logic. :thinking: 

@encode/maintainers If someone is against this idea, or has any concerns regarding this, or want more context, anything... Please let us know. Otherwise, I'll get this merged end of next week.

**Comment by alex-oleshkevich** [comment]
This is useful as it unlocks future per-route error handlers, greatly simplifying view code.
```python
Route('/, view, error_handlers={})
```
However, this does not solve the problem when someone wants to access `request.body` in other middleware. This issue is still there, isn't it? This solution (https://github.com/encode/starlette/pull/1519) is more straightforward as fixes the issue everywhere.

**Comment by adriangb** [comment]
> this does not solve the problem when someone wants to access `request.body` in other middleware. This issue is still there, isn't it? This solution (#1519) is more straightforward as fixes the issue everywhere.

https://github.com/encode/starlette/pull/1692 also resolves it in a way that compliments this PR, and without breaking other things.

**Comment by adriangb** [comment]
Folks, it would be great to get some more input/review here.

File: starlette/middleware/exceptions/_wrapper.py (hunk 1 above)
Comment (adriangb) [reply starlette/middleware/exceptions/_wrapper.py]: Thanks, done in 3cf3881

**Comment by Kludex** [comment]
The last bits here:

What's our plan regarding the `ExceptionMiddleware`? Are we going to remove it at some point?

Also, should we update the documentation: https://www.starlette.io/exceptions/#errors-and-handled-exceptions ?

**Review by Kludex — COMMENTED** [review]
Do you mind if we change the structure of the files around (I can change it)?

I'd prefer to not remove the file `exceptions.py`, and just create a top level `_exception_handler.py`.

**Comment by adriangb** [comment]
Sure go for it

**Comment by Kludex** [comment]
@adriangb I didn't change the documentation, but I think we should do something there for 1.0.

Also, I applied this commit: [`3fb1118` (#2026)](https://github.com/encode/starlette/pull/2026/commits/3fb111835755c65a4144603c552d9d06c54d80e2). If you didn't like it, let me know, so I can revert it. 👍

**Review by Kludex — APPROVED** [review]

**Pull request merged**